File: apps/mobile-native/ios/NexaraApp/Data/Celebraciones.swift
import Foundation

// Cumpleaños y aniversarios de ingreso del equipo para el aviso de Actividades.
// Espejo de `GET me/celebraciones/hoy` (`celebrations.service.ts`) y de los textos de
// `apps/api/src/celebrations/celebraciones.ts`. La edad nunca se muestra: en
// cumpleaños el API ni siquiera la manda (`anios` es nulo).

enum CelebracionTipo {
    static let cumpleanos = "cumpleanos"
    static let aniversario = "aniversario"
}

struct CelebracionHoy: Decodable, Identifiable, Hashable {
    let userId: Int
    let nombre: String
    let avatarUrl: String?
    /// `cumpleanos` | `aniversario`.
    let tipo: String
    /// Solo en aniversarios: años en NEXARA.
    let anios: Int?
    let soyYo: Bool

    var id: String { "\(tipo)-\(userId)" }
    var esCumpleanos: Bool { tipo.lowercased() == CelebracionTipo.cumpleanos }

    private enum CodingKeys: String, CodingKey {
        case userId, nombre, avatarUrl, tipo, anios, soyYo
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        userId = try container.decode(Int.self, forKey: .userId)
        nombre = ((try? container.decode(String.self, forKey: .nombre)) ?? "").trimmingCharacters(in: .whitespacesAndNewlines)
        avatarUrl = try? container.decode(String.self, forKey: .avatarUrl)
        tipo = (try? container.decode(String.self, forKey: .tipo)) ?? CelebracionTipo.cumpleanos
        anios = try? container.decode(Int.self, forKey: .anios)
        soyYo = (try? container.decode(Bool.self, forKey: .soyYo)) ?? false
    }

    /// «Carolina», no «Carolina Juárez Álvarez» (`primerNombre` del API).
    static func primerNombre(_ nombre: String) -> String {
        guard let primero = nombre.split(whereSeparator: { $0.isWhitespace }).first else { return "" }
        return String(primero.prefix(1)).uppercased() + String(primero.dropFirst())
    }

    /// Renglón principal del aviso.
    var titulo: String {
        if esCumpleanos {
            if soyYo {
                let quien = Self.primerNombre(nombre)
                return quien.isEmpty ? "¡Feliz cumpleaños! 🎂" : "¡Feliz cumpleaños, \(quien)! 🎂"
            }
            return "Hoy es cumpleaños de \(nombre.isEmpty ? "alguien del equipo" : nombre) 🎂"
        }
        let n = max(1, anios ?? 1)
        let tiempo = n == 1 ? "1 año" : "\(n) años"
        return "\(nombre.isEmpty ? "Alguien del equipo" : nombre) cumple \(tiempo) en NEXARA 🎉"
    }

    /// Renglón secundario, sin edad.
    var detalle: String {
        if esCumpleanos {
            return soyYo ? "Todo el equipo de NEXARA te desea un gran día." : "Mándale una felicitación."
        }
        return soyYo ? "Gracias por todo lo que aportas al equipo." : "Felicítale por su aniversario."
    }
}

struct CelebracionesHoy: Decodable {
    /// Día de la empresa, `AAAA-MM-DD`: con él se recuerda si ya se cerró el aviso.
    let fecha: String
    let celebraciones: [CelebracionHoy]

    private enum CodingKeys: String, CodingKey {
        case fecha, celebraciones
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        fecha = (try? container.decode(String.self, forKey: .fecha)) ?? ""
        celebraciones = (try? container.decode([CelebracionHoy].self, forKey: .celebraciones)) ?? []
    }
}

/// «Cerrar» vale solo por ese día: mañana, con otra fecha, el aviso vuelve a salir.
enum CelebracionesAviso {
    private static func key(userId: String?) -> String {
        "nx-celebraciones-cerrado-\(userId ?? "anon")"
    }

    static func cerrado(fecha: String, userId: String?) -> Bool {
        !fecha.isEmpty && UserDefaults.standard.string(forKey: key(userId: userId)) == fecha
    }

    static func cerrar(fecha: String, userId: String?) {
        guard !fecha.isEmpty else { return }
        UserDefaults.standard.set(fecha, forKey: key(userId: userId))
    }
}

final class CelebracionesRepository {
    static let shared = CelebracionesRepository()
    private let api = ApiClient.shared
    private init() {}

    /// `GET me/celebraciones/hoy`.
    func deHoy() async throws -> CelebracionesHoy {
        let data = try await api.get("me/celebraciones/hoy")
        return try JSONDecoder().decode(CelebracionesHoy.self, from: data)
    }
}

File: apps/mobile-native/android/app/src/main/java/mx/nexara/mobile/nativeapp/ui/console/activities/Celebraciones.kt
package mx.nexara.mobile.nativeapp.ui.console.activities

import mx.nexara.mobile.nativeapp.data.api.CelebracionDto

/**
 * Textos del aviso de cumpleaños y aniversarios (arriba de Actividades). Puro,
 * sin Android. Mismo tono que los push de `apps/api/src/celebrations/celebraciones.ts`
 * y la misma regla: la edad nunca se menciona.
 */
object Celebraciones {
    const val CUMPLEANOS = "cumpleanos"
    const val ANIVERSARIO = "aniversario"

    fun esCumpleanos(c: CelebracionDto): Boolean = c.tipo?.trim()?.lowercase() == CUMPLEANOS

    fun esAniversario(c: CelebracionDto): Boolean = c.tipo?.trim()?.lowercase() == ANIVERSARIO

    /** Solo lo que la app sabe contar; un tipo nuevo del servidor no sale en blanco. */
    fun visibles(lista: List<CelebracionDto>?): List<CelebracionDto> =
        lista.orEmpty()
            .filter { esCumpleanos(it) || esAniversario(it) }
            // Lo mío primero: es a mí a quien felicitan.
            .sortedBy { if (it.soyYo == true) 0 else 1 }

    /** «Carolina», no «carolina juárez»: primera palabra con mayúscula inicial. */
    fun primerNombre(nombre: String?): String {
        val limpio = nombre?.trim()?.split(Regex("\\s+"))?.firstOrNull().orEmpty()
        return limpio.replaceFirstChar { it.uppercaseChar() }
    }

    /** «Carolina Juárez»: nombre y primer apellido, para no desbordar la tarjeta. */
    fun nombreCorto(nombre: String?): String =
        CoreActivityRules.shortName(nombre).ifBlank { "Alguien del equipo" }

    private fun tiempo(anios: Int): String = if (anios == 1) "1 año" else "$anios años"

    /** Una línea por celebración. Nunca lleva la edad: en cumpleaños `anios` se ignora. */
    fun linea(c: CelebracionDto): String {
        val yo = c.soyYo == true
        return when {
            esCumpleanos(c) && yo -> {
                val quien = primerNombre(c.nombre)
                val saludo = if (quien.isBlank()) "¡Feliz cumpleaños!" else "¡Feliz cumpleaños, $quien!"
                "$saludo 🎂 Todo el equipo te desea un gran día"
            }
            esCumpleanos(c) -> "Hoy es cumpleaños de ${nombreCorto(c.nombre)} 🎂"
            yo -> {
                val quien = primerNombre(c.nombre)
                val saludo = if (quien.isBlank()) "¡Felicidades!" else "¡Felicidades, $quien!"
                val anios = c.anios?.takeIf { it > 0 }
                if (anios != null) "$saludo 🎉 Hoy cumples ${tiempo(anios)} en NEXARA" else "$saludo 🎉 Hoy es tu aniversario en NEXARA"
            }
            else -> {
                val anios = c.anios?.takeIf { it > 0 }
                if (anios != null) {
                    "${nombreCorto(c.nombre)} cumple ${tiempo(anios)} en NEXARA 🎉"
                } else {
                    "Hoy es aniversario de ${nombreCorto(c.nombre)} en NEXARA 🎉"
                }
            }
        }
    }

    /** Texto corto de la fila horizontal: «Ana López 🎂», «Pedro Ruiz · 3 años 🎉». */
    fun etiqueta(c: CelebracionDto): String {
        val quien = if (c.soyYo == true) "Tú" else nombreCorto(c.nombre)
        if (esCumpleanos(c)) return "$quien 🎂"
        val anios = c.anios?.takeIf { it > 0 }
        return if (anios != null) "$quien · ${tiempo(anios)} en NEXARA 🎉" else "$quien 🎉"
    }

    /** El glifo del aviso: pastel si todo es cumpleaños; si hay aniversarios, confeti. */
    fun soloCumpleanos(lista: List<CelebracionDto>): Boolean {
        val principal = visibles(lista)
        val considerar = if (principal.firstOrNull()?.soyYo == true) principal.take(1) else principal
        return considerar.isNotEmpty() && considerar.all { esCumpleanos(it) }
    }

    /** Encabezado cuando hay varias: la mía manda; si no, un título de grupo. */
    fun titulo(lista: List<CelebracionDto>): String? {
        val visibles = visibles(lista)
        return when {
            visibles.isEmpty() -> null
            visibles.size == 1 -> linea(visibles.first())
            visibles.first().soyYo == true -> linea(visibles.first())
            else -> "Hoy celebramos en el equipo 🎉"
        }
    }

    /** Las que van en la fila horizontal debajo del título (todas menos la del título). */
    fun enFila(lista: List<CelebracionDto>): List<CelebracionDto> {
        val visibles = visibles(lista)
        return when {
            visibles.size <= 1 -> emptyList()
            visibles.first().soyYo == true -> visibles.drop(1)
            else -> visibles
        }
    }

    /** Cerrado con la X: se recuerda la fecha, así mañana vuelve a salir. */
    fun cerradoHoy(fechaCerrada: String?, fecha: String?): Boolean =
        !fecha.isNullOrBlank() && fechaCerrada == fecha
}
